compiler/rustc_lint/src/non_fmt_panic.rs

```diff
@@ -69,23 +69,62 @@ fn check_panic<'tcx>(cx: &LateContext<'tcx>, f: &'tcx hir::Expr<'tcx>, arg: &'tc
 
     let (span, panic) = panic_call(cx, f);
 
-    cx.struct_span_lint(NON_FMT_PANIC, arg.span, |lint| {
+    // Find the span of the argument to `panic!()`, before expansion in the
+    // case of `panic!(some_macro!())`.
+    let mut arg_span = arg.span;
+    let mut arg_macro = None;
+    while !span.contains(arg_span) {
+        let expn = arg_span.ctxt().outer_expn_data();
+        if expn.is_root() {
+            break;
+        }
+        arg_macro = expn.macro_def_id;
+        arg_span = expn.call_site;
+    }
```

Comment:
Should this logic be part of `panic_call` so it also applies to `panic!("{}")`?

Reply:
Yes and no. It's a tricky one, because in `fancy_panic!("abc {}")` i still want to point at the `{}` in that string literal, even though `panic!(..)` in the definition of `fancy_panic!()` contained `$x` or something.

There's definitely some improvements that can be made for that situation, but I rather do that in a separate PR, as it's far from trivial what to do in different situations there. :(